transformer_contextual_lenses.py

```diff
@@ -0,0 +1,706 @@
+
+rep_size = 256                 # Size of learned linear representation
+
+# Init PRNG Stream.
+rng = random.PRNGKey(random_seed)
+rng, init_rng = random.split(rng)
+# We init the first set of dropout PRNG keys, but update it afterwards inside
+# the main pmap'd training update for performance.
+dropout_rngs = random.split(rng, jax.local_device_count())
+
+"""# Transformer language model
+Code source: https://github.com/google/flax/blob/master/examples/lm1b/models.py
+"""
+
+class TransformerLM(nn.Module):
+  """Transformer Model for language modeling."""
+
+  def apply(self,
+            inputs,
+            vocab_size,
+            emb_dim=512,
+            num_heads=8,
+            num_layers=6,
+            qkv_dim=512,
+            mlp_dim=2048,
+            max_len=2048,
+            train=False,
+            shift=True,
+            dropout_rate=0.1,
+            attention_dropout_rate=0.1,
+            cache=None):
+    """Applies Transformer model on the inputs.
+    Args:
+      inputs: input data
+      vocab_size: size of the vocabulary
+      emb_dim: dimension of embedding
+      num_heads: number of heads
+      num_layers: number of layers
+      qkv_dim: dimension of the query/key/value
+      mlp_dim: dimension of the mlp on top of attention block
+      max_len: maximum length.
+      train: bool: if model is training.
+      shift: bool: if we right-shift input - this is only disabled for
+        fast, looped single-token autoregressive decoding.
+      dropout_rate: dropout rate
+      attention_dropout_rate: dropout rate for attention weights
+      cache: flax autoregressive cache for fast decoding.
+    Returns:
+      output of a transformer decoder.
+    """
+    padding_mask = jnp.where(inputs > 0, 1, 0).astype(jnp.float32)[..., None]
+    assert inputs.ndim == 2  # (batch, len)
+    x = inputs
+    if shift:
+      x = shift_right(x)
+    x = x.astype('int32')
+
+    x = Embed(x, num_embeddings=vocab_size, features=emb_dim, name='embed')
+
+    x = AddPositionEmbs(
+        x, max_len=max_len, posemb_init=sinusoidal_init(max_len=max_len),
+        cache=cache)
+
+    x = nn.dropout(x, rate=dropout_rate, deterministic=not train)
+
+    for _ in range(num_layers):
+      x = Transformer1DBlock(
+          x,
+          qkv_dim=qkv_dim,
+          mlp_dim=mlp_dim,
+          num_heads=num_heads,
+          causal_mask=True,
+          padding_mask=padding_mask,
+          dropout_rate=dropout_rate,
+          attention_dropout_rate=attention_dropout_rate,
+          deterministic=not train,
+          cache=cache,
+      )
+
+    x = nn.LayerNorm(x)
+
+    logits = nn.Dense(
+        x,
+        vocab_size,
+        kernel_init=nn.initializers.xavier_uniform(),
+        bias_init=nn.initializers.normal(stddev=1e-6))
+
+    return logits
+
+
+
+"""# Transformer lenses
+Based on: https://arxiv.org/pdf/2002.08866.pdf
+
+### Lens 1: Pooling
+
+Mean pooling
+"""
+
+class TransformerMeanPool(nn.Module):
```

Comment:
Instead of making each of these a separate class, how about adding a `reduce_fn` argument to TransformerLM. Can then add
```
if reduce_fn is not None:
  x = reduce_fn(x)
```

Furthermore, can make reduce_fn act on (x, padding_mask), to allow it to ignore padding positions.


```
if reduce_fn is not None:
  x = reduce_fn(x, padding_mask)
```

You can then create the pooling models with:

TransformerMeanPool = TransformerLM.partial(reduce_fn=mean_reduce_fn)

etc